android/src/main/java/it/innove/Peripheral.java

```diff
@@ -220,13 +221,12 @@ public WritableMap asWritableMap(BluetoothGatt gatt) {
         WritableArray characteristicsArray = Arguments.createArray();
 
         if (connected && gatt != null) {
-            for (Iterator<BluetoothGattService> it = gatt.getServices().iterator(); it.hasNext();) {
+            for (Iterator<BluetoothGattService> it = new CopyOnWriteArrayList<>(gatt.getServices()).iterator(); it.hasNext();) {
```

Comment:
I'm doubtful that this will actually fix the problem, but merely hide it.

Reply:
At least will solve for iterator .next()

Reply:
Would you suggest a better approach on that?

Reply:
I would really be interested in the actual reason why the `ConcurrentModificationException` is thrown. 
Since we (i.e. this code) isn't modifying the iterator (unless we call the `next` call), it must mean that the returned `ArrayList` is not a fresh copy, but a reference that gets modified by someone else while we're iterating on it.

I tried (briefly) to find the source code for `BluetoothGatt` but could only manage to find a source that was incorrect (based on the file length and the reported line number on the stack trace in #1229).

If figuring that out is out of scope for this fix (which is completely understandable), I would rather just take a complete copy of the returned array (instead of trying to be fancy with `CopyOnWriteArrayList`).

Reply:
> I would really be interested in the actual reason why the `ConcurrentModificationException` is thrown. Since we (i.e. this code) isn't modifying the iterator (unless we call the `next` call), it must mean that the returned `ArrayList` is not a fresh copy, but a reference that gets modified by someone else while we're iterating on it.
> 
> I tried (briefly) to find the source code for `BluetoothGatt` but could only manage to find a source that was incorrect (based on the file length and the reported line number on the stack trace in #1229).
> 
> If figuring that out is out of scope for this fix (which is completely understandable), I would rather just take a complete copy of the returned array (instead of trying to be fancy with `CopyOnWriteArrayList`).

My guess is that somehow the number of services and characteristics can change during runtime and alongside of asWritableMap is being called.

I agree, do you wanna make that change or i can just abort this pull request and create another one with straight copy of services and characteristics list?

Reply:
I don't have the bandwidth to work on this (nor do I have any permissions to merge anything).